### PR TITLE
Robuster login

### LINKED ISSUE #154: Logging in from options should update the tabs

### PR DESCRIPTION
**Pull request by braxtonhall**
- [x] resolves #154 
- [x] logging out should be supported
    - [x] Remove validator on log out

### CODE REVIEW

File: extension/ts/background/util/dispatchEvent.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+const sendToTab = (event: BackgroundEventMessage) => (tab: Tab) => chrome.tabs.sendMessage(tab.id, event);
+
+const sendToTabs = (event: BackgroundEventMessage) => (tabs: Tab[]) => tabs.forEach(sendToTab(event));
+
+const sendToBackgroundThreads = (event: BackgroundEventMessage) => chrome.runtime.sendMessage(event);
+
+const dispatchEvent = (event: BackgroundEvent) => {
+	const matches = chrome.runtime.getManifest().content_scripts.flatMap((contentScript) => contentScript.matches);
+	const message: BackgroundEventMessage = {kind: event, message: "background-event"};
+	chrome.tabs.query({url: matches}, sendToTabs(message));
```
Comment (adirar111) [inline extension/ts/background/util/dispatchEvent.ts:13]: this makes it less ugly but i know what you are 🫵

File: extension/ts/background/workers/authorize.ts [hunk 2]
```diff
@@ -8,8 +10,12 @@ const authorize = (interactive: AuthorizeParameters): Promise<AuthorizeResponse>
 		if (chrome?.identity?.getAuthToken) {
 			return chrome.identity.getAuthToken({interactive}, (auth) => {
 				if (auth) {
+					if (interactive) {
+						dispatchEvent(BackgroundEvent.CompletedAuth);
+					}
 					resolve(auth);
 				} else {
+					// TODO check runtime.lastError
```
Comment (adirar111) [inline extension/ts/background/workers/authorize.ts:18]: is this still erroring out sometimes

File: extension/ts/main/extensions/tags/index.ts [hunk 3]
```diff
@@ -1,40 +1,47 @@
...
+			container: textAreaContainer,
+			description: "Log in for tag validation and to sync this book's tags with its authors' tags",
+			onLogIn: () => {
+				showFinder();
+				showTagValidator();
+			},
+			onLogOut: () => {
+				hideFinder();
+				hideTagValidator();
 			},
```
Comment (adirar111) [inline extension/ts/main/extensions/tags/index.ts:44]: epic

File: extension/ts/main/extensions/author/authorPage/authorUI.ts [hunk 4]
```diff
@@ -50,44 +50,52 @@ const createEditTagsSection = ({onSave, onPull, onCancel}: ButtonHandlers) => {
...
 		},
-		container,
-		"Log in to manage this author's book tags"
-	);
+		onLogOut: () => {
+			onPushButton.remove();
+			onSyncButton.remove();
+			onEditButton.remove();
+		},
+	});
```
Comment (adirar111) [inline extension/ts/main/extensions/author/authorPage/authorUI.ts:82]: if only there was a multiple remove like the append above lol

**Review by adirar111 — APPROVED** [review]
that moment when home made messaging infrastructure

File: extension/ts/main/extensions/author/authorPage/authorUI.ts (hunk 4 above)
Comment (braxtonhall) [reply extension/ts/main/extensions/author/authorPage/authorUI.ts]: don't let ur dreams be dreams. pushed something but i'm passing out

will check when back from seattle.

or you can try and just merge it if it works

**Pull request merged**